worker/draft.go

```diff
@@ -72,69 +69,102 @@ type node struct {
 	pendingSize int64
 }
 
-type operation struct {
-	// id of the operation.
-	id int
-	// closer is used to signal and wait for the operation to finish/cancel.
-	closer *y.Closer
+type op int
+
+func (id op) String() string {
+	switch id {
+	case opRollup:
+		return "opRollup"
+	case opSnapshot:
+		return "opSnapshot"
+	case opIndexing:
+		return "opIndexing"
+	default:
+		return "opUnknown"
+	}
 }
 
 const (
-	opRollup = iota + 1
+	opRollup op = iota + 1
 	opSnapshot
 	opIndexing
 )
 
 // startTask is used to check whether an op is already going on.
 // If rollup is going on, we cancel and wait for rollup to complete
 // before we return. If the same task is already going, we return error.
-func (n *node) startTask(id int) (*operation, error) {
+func (n *node) startTask(id op) (*y.Closer, error) {
 	n.opsLock.Lock()
 	defer n.opsLock.Unlock()
 
+	stopTask := func(id op) {
+		n.opsLock.Lock()
+		delete(n.ops, id)
+		n.opsLock.Unlock()
+		glog.Infof("Operation completed with id: %s", id)
+
+		// If we were doing any other operation, let's restart rollups.
+		if id != opRollup {
+			time.Sleep(10 * time.Second) // Wait for 10s to start rollup operation.
+			n.startTask(opRollup)
+		}
+	}
+
+	closer := y.NewCloser(1)
 	switch id {
 	case opRollup:
 		if len(n.ops) > 0 {
-			return nil, errors.Errorf("another operation is already running, ops:%v", n.ops)
+			return nil, errors.Errorf("another operation is already running")
 		}
+		go posting.IncrRollup.Process(closer)
+
 	case opSnapshot, opIndexing:
-		if op, has := n.ops[opRollup]; has {
-			glog.Info("Found a rollup going on. Cancelling rollup!")
-			op.closer.SignalAndWait()
-			glog.Info("Rollup cancelled.")
-		} else if len(n.ops) > 0 {
-			return nil, errors.Errorf("another operation is already running, ops:%v", n.ops)
+		for otherId, otherCloser := range n.ops {
+			if otherId == opRollup {
+				otherCloser.SignalAndWait()
```

Comment:
Shouldn't we pause rollup only when we know for sure there are no other operation. Let's say, 
n.ops has opRollup and opIndexing. Then opRollup will be stopped but since there is also an opIndexing, we will return from next line having paused opRollup.